src/completion_provider.rs

```diff
@@ -0,0 +1,335 @@
+    let ret_ptr = match ret {
+        Some(ref ret) => ret.as_ptr(),
+        None => std::ptr::null_mut(),
+    };
+
+    let old_ptr = gobject_sys::g_object_get_qdata(
+        completion_provider as *mut _,
+        COMPLETION_PROVIDER_GICON_QUARK.to_glib(),
+    );
+    if !old_ptr.is_null() {
+        assert_eq!(
+            old_ptr as *mut _, ret_ptr,
+            "Did not return same gicon again"
+        );
+    }
+
+    gobject_sys::g_object_set_qdata_full(
+        completion_provider as *mut _,
+        COMPLETION_PROVIDER_GICON_QUARK.to_glib(),
+        ret_ptr as *mut libc::c_void,
+        Some(unref),
+    );
+
+    ret.to_glib_none().0
+}
+
+unsafe extern "C" fn completion_provider_populate<T: ObjectSubclass + CompletionProviderImpl>(
+    completion_provider: *mut gtk_source_sys::GtkSourceCompletionProvider,
+    context: *mut gtk_source_sys::GtkSourceCompletionContext,
+) {
+    let instance = &*(completion_provider as *mut T::Instance);
+    let imp = instance.get_impl();
+    imp.populate(
+        &from_glib_borrow(completion_provider),
+        &from_glib_borrow(context),
+    )
+}
+
+unsafe extern "C" fn completion_provider_get_activation<
+    T: ObjectSubclass + CompletionProviderImpl,
+>(
+    completion_provider: *mut gtk_source_sys::GtkSourceCompletionProvider,
+) -> gtk_source_sys::GtkSourceCompletionActivation {
+    let instance = &*(completion_provider as *mut T::Instance);
+    let imp = instance.get_impl();
+    imp.get_activation(&from_glib_borrow(completion_provider))
+        .to_glib()
+}
+
+unsafe extern "C" fn completion_provider_match<T: ObjectSubclass + CompletionProviderImpl>(
+    completion_provider: *mut gtk_source_sys::GtkSourceCompletionProvider,
+    context: *mut gtk_source_sys::GtkSourceCompletionContext,
+) -> glib_sys::gboolean {
+    let instance = &*(completion_provider as *mut T::Instance);
+    let imp = instance.get_impl();
+    imp.provide_match(
+        &from_glib_borrow(completion_provider),
+        &from_glib_borrow(context),
+    )
+    .to_glib()
+}
+
+static COMPLETION_PROVIDER_INFO_WIDGET: Lazy<glib::Quark> =
+    Lazy::new(|| glib::Quark::from_string("gtk-sourceview-subclass-info-widget"));
+
+unsafe extern "C" fn completion_provider_get_info_widget<
+    T: ObjectSubclass + CompletionProviderImpl,
+>(
+    completion_provider: *mut gtk_source_sys::GtkSourceCompletionProvider,
+    proposal: *mut gtk_source_sys::GtkSourceCompletionProposal,
+) -> *mut gtk_sys::GtkWidget {
+    let instance = &*(completion_provider as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    let ret = imp.get_info_widget(
+        &from_glib_borrow(completion_provider),
+        &from_glib_borrow(proposal),
+    );
+    let ret_ptr = match ret {
+        Some(ref ret) => ret.as_ptr(),
+        None => std::ptr::null_mut(),
+    };
+
+    let old_ptr = gobject_sys::g_object_get_qdata(
+        completion_provider as *mut _,
```

Comment:
As we discussed this widget can be different for each proposal. So forbidding this seems like a problem, and the lifetime should probably be bound (via the qdata here) to the proposal and not to the provider.